android/java/org/chromium/chrome/browser/settings/BraveNewsPreferences.java

```diff
@@ -0,0 +1,217 @@
+    public static final String PREF_SHOW_NEWS = "kNewTabPageShowToday";
+    public static final String PREF_SHOW_OPTIN = "show_optin";
+    public static final String PREF_SOURCES_SECTION = "your_sources_section";
+    public static final String PREF_ADD_SOURCES = "add_source_news";
+
+    private ChromeSwitchPreference turnOnNews;
+    private ChromeSwitchPreference showNews;
+    private PreferenceCategory yourSources;
+    private EditText mEditText;
+    private PreferenceScreen mainScreen;
+    private PreferenceManager preferenceManager;
+    private TreeMap<String, List<Publisher>> categsPublishers;
+
+    private final HashMap<String, Preference> mRemovedPreferences = new HashMap<>();
+    private BraveNewsController mBraveNewsController;
+    private PreferenceFragmentCompat settingsFragment;
+
+    public static int getPreferenceSummary() {
+        return BraveLaunchIntentDispatcher.useCustomTabs() ? R.string.text_on : R.string.text_off;
+    }
+
+    @Override
+    public void onCreatePreferences(@Nullable Bundle savedInstanceState, String rootKey) {
+        SettingsUtils.addPreferencesFromResource(this, R.xml.brave_news_preferences);
+        InitBraveNewsController();
+        turnOnNews = (ChromeSwitchPreference) findPreference(PREF_TURN_ON_NEWS);
+        showNews = (ChromeSwitchPreference) findPreference(PREF_SHOW_NEWS);
+        settingsFragment = this;
+
+        turnOnNews.setOnPreferenceChangeListener(this);
+        showNews.setOnPreferenceChangeListener(this);
+
+        categsPublishers = new TreeMap<>();
+        mBraveNewsController.getPublishers((publishers) -> {
+            List<Publisher> allPublishers = new ArrayList<>();
+            List<Publisher> categoryPublishers = new ArrayList<>();
+            for (Map.Entry<String, Publisher> entry : publishers.entrySet()) {
+                String key = entry.getKey();
+                Publisher publisher = entry.getValue();
+                // categsPublishers.computeIfAbsent(publisher.categoryName, k ->new
+                // ArrayList<>()).add(publisher);
+                categoryPublishers.add(publisher);
+                categsPublishers.put(publisher.categoryName, categoryPublishers);
+            }
+            categsPublishers.put("All Sources", allPublishers);
+            addCategs(categsPublishers);
+        });
+    }
+
+    private void addCategs(TreeMap<String, List<Publisher>> publisherCategories) {
+        for (Map.Entry<String, List<Publisher>> map : publisherCategories.entrySet()) {
+            String category = map.getKey();
+            List<Publisher> publishers = map.getValue();
+
+            ChromeBasePreference source =
+                    new ChromeBasePreference(ContextUtils.getApplicationContext());
+            source.setTitle(category);
+            source.setKey(category);
+            Bundle prefExtras = source.getExtras();
+            prefExtras.putString("category", category);
```

Comment:
after putting this mapping into bundle `prefExtras.putString("category", category);`, prefExtras is not passed anywhere. should it be like that?

Reply:
no, it was leftover. removed

Reply:
actually i'll have to revert. it's used to add the category name for the next pref screen

Reply:
sorry, my bad, I didn't understand this block correctly.